eo-maven-plugin/src/main/java/org/eolang/maven/DiscoverMojo.java

```diff
@@ -94,26 +95,33 @@ public void exec() throws IOException {
     private Collection<String> discover(final Path file)
         throws FileNotFoundException {
         final XML xml = new XMLDocument(file);
+        final List<String> base = new ListOf<>(
+            "//o[",
+            "not(starts-with(@base,'.'))",
+            "and @base != 'Q'",
+            "and @base != '^'",
+            "and @base != '$'",
+            "and @base != '&'",
+            "and not(@ref)"
+        );
+        final List<String> regular = new ListOf<>(base);
+        if (this.withVersions) {
```

Comment:
@maxonfjvipon I don't clearly understand why we need that check, to be honest. As I understand, you are looking the names for `regular` and `versioned` objects separately. Do we have `ver` attribute if `this.withVersions==false`? Can we just get `regular` objects (for all cases) as follows:
```java
 final List<String> regular = new ListOf<>(
            "//o[",
            "not(starts-with(@base,'.'))",
            "and @base != 'Q'",
            "and @base != '^'",
            "and @base != '$'",
            "and @base != '&'",
            "and not(@ref)",
            "and not(@ver)"
            "]/@base"
        );
```

Moreover, I would suggest to move the logic with creation of `regular` and `versioned` names into separate methods. For me (personally) it looks more logical.

Reply:
@volodya-lombrozo 
1. Regular. We select `/@base` in both cases (either `withVersions` is `true` or `false`) because at this step (discover) there are can be objects without versions (even if `withVersions` is `true`). Here in `regular` if `withVersions` is `true` we add `and not(@ver)` to exclude objects with `ver` attribute because they will be selected at the step 2.
2. Versioned. If `withVersions` is `true` - we additionally select `/concat(@base,'|',@ver)` for objects with `ver` attribute. I didn't find a way to select conditionally in xpath.

I hope you'll understand what I mean)

Reply:
Let's clarify, then :) 

Imagine, we have the next objects:

```xml
<o ver="1" base="foo"/>
<o base="bar"/>
<o ver="2" base="foobar">
```

And we have two options:
1) `withVersions = true`:

```java
 final List<String> regular = new ListOf<>(
            "//o[",
            "not(starts-with(@base,'.'))",
            "and @base != 'Q'",
            "and @base != '^'",
            "and @base != '$'",
            "and @base != '&'",
            "and not(@ref)",
            "and not(@ver)"
            "]/@base"
        );
```

Res: 
```xml
"bar"
```

```java
 final List<String> versioned = new ListOf<>(
            "//o[",
            "not(starts-with(@base,'.'))",
            "and @base != 'Q'",
            "and @base != '^'",
            "and @base != '$'",
            "and @base != '&'",
            "and not(@ref)",
            "and @ver",
            "]/concat(@base,'|',@ver)"
 );
```

Res:

```xml
"foo|1", "foobar|2"
```

**names="bar, foo|1, foobar|2"**

2) `withVersions=false`:
```java
 final List<String> regular = new ListOf<>(
            "//o[",
            "not(starts-with(@base,'.'))",
            "and @base != 'Q'",
            "and @base != '^'",
            "and @base != '$'",
            "and @base != '&'",
            "and not(@ref)",
            "]/@base"
);
```

Res:
```xml
"foo"
"bar"
"foobar"
```

```java
 final List<String> versioned = NA
```

Res: `empty`

**names = "foo, bar, foobar"**

___

So, suggest:

1. `withVersions = true`:

```java
 final List<String> all = new ListOf<>(
            "//o[",
            "not(starts-with(@base,'.'))",
            "and @base != 'Q'",
            "and @base != '^'",
            "and @base != '$'",
            "and @base != '&'",
            "and not(@ref)"
            "]/string-join((@base, @ver), '|')"
        );
```

**names="bar, foo|1, foobar|2"**

2. `withVersions = false`:

```java
 final List<String> all = new ListOf<>(
            "//o[",
            "not(starts-with(@base,'.'))",
            "and @base != 'Q'",
            "and @base != '^'",
            "and @base != '$'",
            "and @base != '&'",
            "and not(@ref)"
            "]/@base"
        );
```
**names = "foo, bar, foobar"**

And you don’t need `regular` and `versioned` lists.

Reply:
@maxonfjvipon

Reply:
@volodya-lombrozo if `/string-join((@base,@ver),'|')` returns the same as `/@base` if `@ver` does not exist then it can be one unique xpath for both cases, but I'm not sure

Reply:
@volodya-lombrozo I was wrong, forget about it

Reply:
@volodya-lombrozo so did some research, you suggestion won't work.
```java
 final List<String> all = new ListOf<>(
    "//o[",
    "not(starts-with(@base,'.'))",
    "and @base != 'Q'",
    "and @base != '^'",
    "and @base != '$'",
    "and @base != '&'",
    "and not(@ref)"
    "]/string-join((@base, @ver), '|')"
);
```
If there are some object where `@ver` is not present - names won't be found, so we need to check if `@ver` is present or not before getting the result

Reply:
@maxonfjvipon Ok, let's discuss it in a separate puzzle then.